src/iface/ethernet.rs

```diff
@@ -759,11 +732,146 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
+    ///         .ethernet_addr(hw_addr)
+    ///         .neighbor_cache(neighbor_cache)
+    ///         .ip_addrs(ip_addrs)
+    ///         .finalize();
+    /// ```
+    pub fn new(device: DeviceT) -> InterfaceBuilder<'b, 'c, DeviceT> {
+        let caps = device.capabilities();
+        InterfaceBuilder {
+            device:              device,
+            device_capabilities: caps,
+            ethernet_addr:       None,
+            neighbor_cache:      None,
+            ip_addrs:            None,
+            ipv4_gateway:        None
+        }
+    }
+
+    /// Set the Ethernet address the interface will use. See also
+    /// [ethernet_addr].
+    ///
+    /// # Panics
+    /// This function panics if the address is not unicast.
+    ///
+    /// [ethernet_addr]: struct.EthernetInterface.html#method.ethernet_addr
+    pub fn ethernet_addr(mut self, addr: EthernetAddress) -> InterfaceBuilder<'b, 'c, DeviceT> {
+        InterfaceInner::check_ethernet_addr(&addr);
+        self.ethernet_addr = Some(addr);
+        self
+    }
+
+    /// Set the IP addresses the interface will use. See also
+    /// [ip_addrs].
+    ///
+    /// # Panics
+    /// This function panics if any of the addresses is not unicast.
+    ///
+    /// [ip_addrs]: struct.EthernetInterface.html#method.ip_addrs
+    pub fn ip_addrs<T: Into<ManagedSlice<'c, IpCidr>>>(mut self, ips: T) -> InterfaceBuilder<'b, 'c, DeviceT> {
+        let ips = ips.into();
+        InterfaceInner::check_ip_addrs(&ips);
+        self.ip_addrs = Some(ips);
+        self
+    }
+
+    /// Set the IPv4 gateway the interface will use. See also
+    /// [ipv4_gateway].
+    ///
+    /// # Panics
+    /// This function panics if the given address is not unicast.
+    ///
+    /// [ipv4_gateway]: struct.EthernetInterface.html#method.ipv4_gateway
+    pub fn ipv4_gateway<T>(mut self, gateway: T) -> InterfaceBuilder<'b, 'c, DeviceT>
+            where T: Into<Ipv4Address> {
+        let addr = gateway.into();
+        if !addr.is_unicast() {
+            panic!("IP address {} is not unicast", addr);
+        }
+        self.ipv4_gateway = Some(addr);
+        self
+    }
+
+    /// Set the Neighbor Cache the interface will use.
+    pub fn neighbor_cache(mut self, neighbor_cache: NeighborCache<'b>) -> InterfaceBuilder<'b, 'c, DeviceT> {
+        self.neighbor_cache = Some(neighbor_cache);
+        self
+    }
+
+    /// Create a network interface using the previously provided configuration.
+    ///
+    /// # Panics
+    /// If a required option is not provided, this function will panic. Required
+    /// options include.
+    ///
+    /// - [ethernet_addr]
+    /// - [neighbor_cache]
+    /// - [ip_addrs]
+    ///
+    /// [ethernet_addr]: #method.ethernet_addr
+    /// [neighbor_cache]: #method.neighbor_cache
+    /// [ip_addrs]: #method.ip_addrs
+    pub fn finalize(self) -> Interface<'b, 'c, DeviceT> {
+        // TODO: Limit the number of required options.
```

Comment:
What do you mean by this?

Reply:
For example, I wondered if it is maybe possible to `finalize` without `ip_addrs` so that.

```rust
EthernetInterfaceBuilder::new(Loopback::new())
        .ip_addrs([])
        ...
```

Becomes

```rust
EthernetInterfaceBuilder::new(Loopback::new())
        ...
```

Reply:
Ah sure, that works.